src/_guides/language/effective-dart/design_migrated.md

```diff
@@ -1218,90 +1289,167 @@ var names = people.map((person) => person.name);
+variable is *not* a local then instead of writing the type argument list on the
+invocation itself, you may put a type annotation on the declaration:
+
+{:.good}
+<?code-excerpt "design_good.dart (inferred-type-args)"?>
+{% prettify dart tag=pre+code %}
+class Downloader {
+  final Completer<String> response = Completer();
+}
+{% endprettify %}
+
+{:.bad}
+<?code-excerpt "design_bad.dart (inferred-type-args)"?>
+{% prettify dart tag=pre+code %}
+class Downloader {
+  final response = Completer();
+}
+{% endprettify %}
+
+Annotating the variable also addresses this guideline because now the type
+arguments *are* inferred.
+
+
+### DON'T write type arguments on generic invocations that are inferred.
+
+This is the converse of the previous rule. If an invocation's type argument list
+*is* correctly inferred with the types you want, then omit the types and let
+Dart do the work for you.
 
 {:.good}
 <?code-excerpt "design_good.dart (redundant)"?>
 {% prettify dart tag=pre+code %}
-Set<String> things = Set();
+class Downloader {
+  final Completer<String> response = Completer();
+}
 {% endprettify %}
 
 {:.bad}
 <?code-excerpt "design_bad.dart (redundant)"?>
 {% prettify dart tag=pre+code %}
-Set<String> things = Set<String>();
+class Downloader {
+  final Completer<String> response = Completer<String>();
+}
 {% endprettify %}
 
-Here, the type annotation on the variable is used to infer the type argument of
-constructor call in the initializer.
-
-In other contexts, there isn't enough information to infer the type and then you
-should write the type argument:
+Here, the type annotation on the field provides a surrounding context to infer
+the type argument of constructor call in the initializer.
 
 {:.good}
 <?code-excerpt "design_good.dart (explicit)"?>
 {% prettify dart tag=pre+code %}
-var things = Set<String>();
+var items = Future.value([1, 2, 3]);
 {% endprettify %}
 
 {:.bad}
 <?code-excerpt "design_bad.dart (explicit)"?>
 {% prettify dart tag=pre+code %}
-var things = Set();
+var items = Future<List<int>>.value(<int>[1, 2, 3]);
 {% endprettify %}
 
-Here, since the variable has no type annotation, there isn't enough context to
-determine what kind of `Set` to create, so the type argument should be provided
-explicitly.
+Here, the types of the collection and instance can be inferred bottom-up from
+their elements and arguments.
 
 
-### DO annotate when Dart infers the wrong type.
+### AVOID writing incomplete generic types.
 
-Sometimes, Dart infers a type, but not the type you want. For example, you may
-want a variable's type to be a supertype of the initializer's type so that you
-can later assign some other sibling type to the variable:
+When writing a type annotation or type argument, it's still possible to not pin
+down a complete type by writing the name of a generic type but omitting its type
+arguments. In Java, these are called "raw types". For example:
```

Comment:
I'm having a hard time parsing the first sentence. Could you please rewrite it?

Reply:
Done!